src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/EETypeNode.cs

```diff
@@ -94,25 +127,155 @@ public override ObjectData GetData(NodeFactory factory, bool relocsOnly)
-                objData.EmitShort(0x4);                           // m_flags: IsArray(0x4)
+                objData.EmitShort((short)((ArrayType)_type).ElementType.GetElementSize());
             }
             else if (_type.IsString)
             {
-                objData.EmitShort(2); // m_ComponentSize
-                objData.EmitShort(0); // m_flags: 0
+                objData.EmitShort(2);
             }
             else
             {
-                objData.EmitShort(0); // m_ComponentSize
-                objData.EmitShort(0); // m_flags: 0
+                objData.EmitShort(0);
+            }
+        }
+        
+        private void OutputFlags(NodeFactory factory, ref ObjectDataBuilder objData)
+        {
+            // Todo: RelatedTypeViaIATFlag when we support cross-module EETypes
+            // Todo: GenericVarianceFlag when we support variance
+            // Todo: OptionalFieldsFlag
+
+            UInt16 flags = (UInt16)EETypeFlags.CanonicalEEType;
+            
+            if (_type.IsArray || _type.IsPointer)
+            {
+                flags = (UInt16)EETypeFlags.ParameterizedEEType;
+            }
+            
+            if (_type.IsValueType)
+            {
+                flags |= (UInt16)EETypeFlags.ValueTypeFlag;
+            }
+
+            if (_type.HasFinalizer)
+            {
+                flags |= (UInt16)EETypeFlags.HasFinalizerFlag;
+            }
+            
+            if (_type is MetadataType && ((MetadataType)_type).ContainsPointers)
+            {
+                flags |= (UInt16)EETypeFlags.HasPointersFlag;
+            }
+            else if (_type.IsArray)
+            {
+                ArrayType arrayType = _type as ArrayType;
+                if ((arrayType.ElementType.IsValueType && arrayType.ElementType is MetadataType && ((MetadataType)arrayType.ElementType).ContainsPointers) ||
+                    !arrayType.ElementType.IsValueType)
+                {
+                    flags |= (UInt16)EETypeFlags.HasPointersFlag;
+                }
+            }
+            
+            if (_type.IsInterface)
+            {
+                flags |= (UInt16)EETypeFlags.IsInterfaceFlag;
+            }
+
+            if (_type.HasInstantiation)
+            {
+                flags |= (UInt16)EETypeFlags.IsGenericFlag;
             }
 
+            int corElementType = 0;
+            
+            // The top 5 bits of flags are used to convey enum underlying type, primitive type, or mark the type as being System.Array
+            if (_type.IsEnum)
+            {
+                TypeDesc underlyingType = _type.UnderlyingType;
+                corElementType = (int)underlyingType.Category;
+                Debug.Assert((int)TypeFlags.SByte <= corElementType && corElementType <= (int)TypeFlags.UInt64);
+            }
+            else if (_type.Category >= TypeFlags.Boolean && _type.Category <= TypeFlags.Double)
+            {
+                corElementType = (int)_type.Category;
```

Comment:
The CorInfoType matches the Category, but ECMA CorElementType does not match neither CorInfoType or Category.

E.g. For double: ECMA CorElementType is 0x0d, but Category is 0x0F.

(The way ECMA CorElementType is defined is lame because of it has holes in it. It would be nice to clean this up once it is ok to make breaking change in MRT file format and use the category enum for EEType too. Maybe we can make this change for CoreRT-only for now.)